src/bin/cargo/commands/help.rs

```diff
@@ -0,0 +1,128 @@
+        .skip(1)
+        .skip_while(|arg| arg.to_str().map_or(false, |s| s.starts_with('-')));
+    if !args
+        .next()
+        .map_or(false, |arg| arg.to_str() == Some("help"))
+    {
+        return Ok(false);
+    }
+    let subcommand = match args.next() {
+        Some(arg) => arg,
+        None => return Ok(false),
+    };
+    let subcommand = match subcommand.to_str() {
+        Some(s) => s,
+        None => return Ok(false),
+    };
+    // Check if this is a built-in command (or alias);
+    let subcommand = match check_alias(config, subcommand) {
+        Some(s) => s,
+        None => return Ok(false),
+    };
+    if resolve_executable(Path::new("man")).is_ok() {
+        let man = match extract_man(&subcommand, "1")? {
+            Some(man) => man,
+            None => return Ok(false),
+        };
+        write_and_spawn(&man, "man")?;
+    } else {
+        let txt = match extract_man(&subcommand, "txt")? {
+            Some(txt) => txt,
+            None => return Ok(false),
+        };
+        if resolve_executable(Path::new("less")).is_ok() {
+            write_and_spawn(&txt, "less")?;
+        } else if resolve_executable(Path::new("more")).is_ok() {
+            write_and_spawn(&txt, "more")?;
+        } else {
+            drop(std::io::stdout().write_all(&txt));
+        }
+    }
+    Ok(true)
+}
+
+/// Checks if the given subcommand is a built-in command (possibly via an alias).
+///
+/// Returns None if it is not a built-in command.
+fn check_alias(config: &Config, subcommand: &str) -> Option<String> {
+    if super::builtin_exec(subcommand).is_some() {
+        return Some(subcommand.to_string());
+    }
+    match aliased_command(config, subcommand) {
+        Ok(Some(alias)) => {
+            let alias = alias.into_iter().next()?;
+            if super::builtin_exec(&alias).is_some() {
+                Some(alias)
+            } else {
+                None
+            }
+        }
+        _ => None,
+    }
+}
+
+/// Extracts the given man page from the compressed archive.
+///
+/// Returns None if the command wasn't found.
+fn extract_man(subcommand: &str, extension: &str) -> CargoResult<Option<Vec<u8>>> {
+    let extract_name = OsString::from(format!("cargo-{}.{}", subcommand, extension));
+    let gz = GzDecoder::new(COMPRESSED_MAN);
+    let mut ar = tar::Archive::new(gz);
+    for entry in ar.entries().unwrap() {
+        let mut entry = entry.unwrap();
+        let path = entry.path().unwrap();
+        if path.file_name().unwrap() != extract_name {
+            continue;
+        }
+        let mut result = Vec::new();
+        entry.read_to_end(&mut result).unwrap();
+        return Ok(Some(result));
+    }
+    Ok(None)
+}
+
+/// Write the contents of a man page to disk and spawn the given command to
+/// display it.
+fn write_and_spawn(contents: &[u8], command: &str) -> CargoResult<()> {
+    let mut tmp = tempfile::Builder::new().prefix("cargo-man").tempfile()?;
```

Comment:
Is the tempfile required here? Programs like `less` I think we can just write everything to a pipe? 

I'm not actually sure what other common CLI apps do here.

Reply:
Unfortunately, AFAIK, the `man` program does not support piped content.